### PR TITLE
feat: Track pod count as a scalar resource

### PR DESCRIPTION
**Pull request by itsomri**
## Description

This PR fixed max pod tracking on scheduling simulations.

Changes:
- "Pods" are now tracked as a scalar resource 
- Added an e2e test to verify preemption simulates pod limits correctly
- Changed node max pods predicate to handle edge cases for reservation pods

## Checklist

- [ ] Self-reviewed
- [ ] Added/updated tests (if needed)
- [ ] Updated documentation (if needed)

### CODE REVIEW

**Comment by itsomri** [comment]
@coderabbitai review

File: pkg/scheduler/api/resource_info/resource_requirment.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ import (
 const (
 	GPUResourceName    = "nvidia.com/gpu"
 	amdGpuResourceName = "amd.com/gpu"
+	PodsResourceName   = v1.ResourceName("pods")
```
Comment (gshaibi) [inline pkg/scheduler/api/resource_info/resource_requirment.go]: Should we use `v1.ResourcePods`?

File: pkg/scheduler/api/resource_info/base_resources.go [hunk 2]
```diff
@@ -55,6 +55,9 @@ func (r *BaseResource) Add(other *BaseResource) {
 	r.memory += other.memory
 	for rName, rrValue := range other.scalarResources {
 		r.scalarResources[rName] += rrValue
+		if r.scalarResources[rName] == 0 {
```
Comment (gshaibi) [inline pkg/scheduler/api/resource_info/base_resources.go:58]: Just out of curiosity - what was the trigger for this change?

File: pkg/scheduler/api/resource_info/resource_requirment.go [hunk 3]
```diff
@@ -153,10 +154,15 @@ func (r *ResourceRequirements) DetailedString() string {
...
 	for rName, rQuant := range r.scalarResources {
-		if rName == v1.ResourceEphemeralStorage || rName == v1.ResourceStorage {
+		switch rName {
+		case v1.ResourceEphemeralStorage, v1.ResourceStorage:
 			rQuant = rQuant / int64(MemoryToGB) // convert from milli-bytes to GB
+			messageBuilder.WriteString(fmt.Sprintf(", %s: %v (GB)", rName, rQuant))
+		case v1.ResourcePods:
+			messageBuilder.WriteString(fmt.Sprintf(", %s: %v", rName, rQuant))
+		default:
+			messageBuilder.WriteString(fmt.Sprintf(", %s: %v", rName, rQuant))
```
Comment (gshaibi) [inline pkg/scheduler/api/resource_info/resource_requirment.go]: If pods and default cases are the same, can we use the default only?

File: pkg/scheduler/gpu_sharing/gpuSharing.go [hunk 4]
```diff
@@ -12,15 +12,15 @@ import (
 	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/log"
 )
 
-type nodeGpuForSharing struct {
+type NodeGpuForSharing struct {
```
Comment (gshaibi) [inline pkg/scheduler/gpu_sharing/gpuSharing.go]: Why do we need it public?

File: pkg/scheduler/plugins/predicates/predicates.go [hunk 5]
```diff
@@ -262,4 +261,69 @@ func (pp *predicatesPlugin) evaluateTaskOnPredicates(
...
+	}
+
+	return nil
+}
+
+// willCreateNewGpuGroup determines if allocating this task will create a new GPU group
+// (and thus require a new reservation pod).
+func (pp *predicatesPlugin) willCreateNewGpuGroup(task *pod_info.PodInfo, node *node_info.NodeInfo) bool {
+	if pp.ssn == nil {
+		return false
```
Comment (gshaibi) [inline pkg/scheduler/plugins/predicates/predicates.go]: Maybe we should return here `true` to be on the safe side?

File: pkg/scheduler/api/resource_info/base_resources.go (hunk 2 above)
Comment (itsomri) [reply pkg/scheduler/api/resource_info/base_resources.go:58]: In tests, there are DeepEqual comparisons that don't see resource=0 the same as not having the resource in the map. I don't really like adding this deletion but couldn't see a better option

File: pkg/scheduler/plugins/predicates/predicates.go (hunk 5 above)
Comment (itsomri) [reply pkg/scheduler/plugins/predicates/predicates.go]: Fixed

File: pkg/scheduler/gpu_sharing/gpuSharing.go (hunk 4 above)
Comment (itsomri) [reply pkg/scheduler/gpu_sharing/gpuSharing.go]: Made private again

File: pkg/scheduler/api/resource_info/resource_requirment.go (hunk 3 above)
Comment (itsomri) [reply pkg/scheduler/api/resource_info/resource_requirment.go]: Fixed

File: pkg/scheduler/api/resource_info/resource_requirment.go (hunk 1 above)
Comment (itsomri) [reply pkg/scheduler/api/resource_info/resource_requirment.go]: done
